File: client_sdks/devconnect-android/src/main/java/com/devconnect/plugins/PerformanceMonitor.kt
package com.devconnect.plugins

import android.app.Activity
import android.app.Application
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.Choreographer
import com.devconnect.DevConnect

private var running = false
private var memoryHandler: Handler? = null
private var memoryRunnable: Runnable? = null
private var cpuHandler: Handler? = null
private var cpuRunnable: Runnable? = null
private var systemHandler: Handler? = null
private var systemRunnable: Runnable? = null
private var startupReported = false
private var appContext: Context? = null
private var lastMemoryMB = 0.0

data class PerformanceMonitorOptions(
    val fpsInterval: Long = 2000L,
    val memoryInterval: Long = 5000L,
    val cpuInterval: Long = 3000L,
    val systemInterval: Long = 10000L,
    val jankThresholdMs: Double = 32.0
)

private val initTimeMs = System.currentTimeMillis()

fun startPerformanceMonitor(context: Any? = null, opts: PerformanceMonitorOptions = PerformanceMonitorOptions()) {
    if (running) return
    running = true

    // Save context for battery/thermal
    if (context is Context) appContext = context.applicationContext
    if (context is Activity) appContext = context.applicationContext

    // ---- Startup Time ----
    if (!startupReported) {
        val startupMs = System.currentTimeMillis() - initTimeMs
        DevConnect.reportPerformanceMetric(
            metricType = "startup_time",
            value = startupMs.toDouble(),
            label = "App startup: ${startupMs}ms"
        )
        startupReported = true
    }

    // ---- FPS + Jank + Frame Timing via Choreographer ----
    // Choreographer.getInstance() requires a thread with a Looper.
    // PerformanceMonitor may be invoked from a worker coroutine
    // (e.g. inside DevConnect.init's IO-dispatched initScope), so
    // post the frame-callback registration to the main thread.
    Handler(Looper.getMainLooper()).post { startFrameMonitor(opts) }

    // ---- Memory monitor ----
    val handler = Handler(Looper.getMainLooper())
    memoryHandler = handler
    val runnable = object : Runnable {
        override fun run() {
            if (!running) return
            reportMemory()
            handler.postDelayed(this, opts.memoryInterval)
        }
    }
    memoryRunnable = runnable
    handler.postDelayed(runnable, opts.memoryInterval)

    // ---- CPU monitor ----
    val cpuH = Handler(Looper.getMainLooper())
    cpuHandler = cpuH
    lastCpuTime = android.os.Process.getElapsedCpuTime()
    lastCpuWallTime = System.currentTimeMillis()
    val cpuR = object : Runnable {
        override fun run() {
            if (!running) return
            reportCpu()
            cpuH.postDelayed(this, opts.cpuInterval)
        }
    }
    cpuRunnable = cpuR
    cpuH.postDelayed(cpuR, opts.cpuInterval)

    // ---- System metrics (battery, threads, disk, thermal) ----
    val sysH = Handler(Looper.getMainLooper())
    systemHandler = sysH
    reportSystemMetrics() // Report immediately
    val sysR = object : Runnable {
        override fun run() {
            if (!running) return
            reportSystemMetrics()
            sysH.postDelayed(this, opts.systemInterval)
        }
    }
    systemRunnable = sysR
    sysH.postDelayed(sysR, opts.systemInterval)
}

fun stopPerformanceMonitor() {
    running = false
    memoryHandler?.removeCallbacksAndMessages(null)
    memoryHandler = null
    memoryRunnable = null
    cpuHandler?.removeCallbacksAndMessages(null)
    cpuHandler = null
    cpuRunnable = null
    systemHandler?.removeCallbacksAndMessages(null)
    systemHandler = null
    systemRunnable = null
}

// ---- Frame monitoring ----

private val frameSampleRate = 5 // report every Nth frame to reduce overhead
private var frameCount = 0
private var lastFpsReportTime = 0L
private var lastFrameTime = 0L
private var lastCpuTime = 0L
private var lastCpuWallTime = 0L

private fun startFrameMonitor(opts: PerformanceMonitorOptions) {
    lastFpsReportTime = System.currentTimeMillis()
    lastFrameTime = System.nanoTime()

    val callback = object : Choreographer.FrameCallback {
        override fun doFrame(frameTimeNanos: Long) {
            if (!running) return

            val now = System.nanoTime()
            if (lastFrameTime > 0) {
                val frameDeltaMs = (now - lastFrameTime) / 1_000_000.0
                frameCount++

                // Report frame build time (sampled)
                if (frameCount % frameSampleRate == 0) {
                    DevConnect.reportPerformanceMetric(
                        metricType = "frame_build_time",
                        value = Math.round(frameDeltaMs * 10.0) / 10.0,
                        label = "Frame: ${Math.round(frameDeltaMs)}ms"
                    )
                }

                // Detect jank
                if (frameDeltaMs > opts.jankThresholdMs) {
                    DevConnect.reportPerformanceMetric(
                        metricType = "jank_frame",
                        value = Math.round(frameDeltaMs * 10.0) / 10.0,
                        label = "Slow frame: ${Math.round(frameDeltaMs)}ms",
                        metadata = mapOf("threshold" to opts.jankThresholdMs)
                    )
                }
            }
            lastFrameTime = now

            // Report FPS periodically
            val nowMs = System.currentTimeMillis()
            val elapsed = nowMs - lastFpsReportTime
            if (elapsed >= opts.fpsInterval) {
                val fps = Math.round(frameCount.toDouble() / elapsed * 1000 * 10) / 10.0
                DevConnect.reportPerformanceMetric(
                    metricType = "fps",
                    value = fps,
                    label = "UI Thread FPS"
                )
                frameCount = 0
                lastFpsReportTime = nowMs
            }

            Choreographer.getInstance().postFrameCallback(this)
        }
    }

    Choreographer.getInstance().postFrameCallback(callback)
}

private fun reportCpu() {
    try {
        val currentCpuTime = android.os.Process.getElapsedCpuTime()
        val currentWallTime = System.currentTimeMillis()
        val cpuDelta = currentCpuTime - lastCpuTime
        val wallDelta = currentWallTime - lastCpuWallTime
        if (wallDelta > 0) {
            val cores = Runtime.getRuntime().availableProcessors()
            val usage = (cpuDelta.toDouble() / wallDelta * 100 / cores).coerceIn(0.0, 100.0)
            DevConnect.reportPerformanceMetric(
                metricType = "cpu_usage",
                value = Math.round(usage * 10.0) / 10.0,
                label = "Process CPU (%)",
                metadata = mapOf(
                    "cpuTimeMs" to cpuDelta,
                    "wallTimeMs" to wallDelta,
                    "cores" to cores
                )
            )
        }
        lastCpuTime = currentCpuTime
        lastCpuWallTime = currentWallTime
    } catch (_: Exception) {}
}

private fun reportMemory() {
    val runtime = Runtime.getRuntime()
    val usedMB = (runtime.totalMemory() - runtime.freeMemory()).toDouble() / 1024 / 1024
    val maxMB = runtime.maxMemory().toDouble() / 1024 / 1024

    DevConnect.reportPerformanceMetric(
        metricType = "memory_usage",
        value = Math.round(usedMB * 10.0) / 10.0,
        label = "JVM Heap Used (MB)",
        metadata = mapOf(
            "totalMemory" to runtime.totalMemory(),
            "freeMemory" to runtime.freeMemory(),
            "maxMemory" to runtime.maxMemory(),
            "maxMB" to Math.round(maxMB * 10.0) / 10.0
        )
    )

    // Native heap
    try {
        val nativeHeap = android.os.Debug.getNativeHeapAllocatedSize().toDouble() / 1024 / 1024
        DevConnect.reportPerformanceMetric(
            metricType = "memory_peak",
            value = Math.round(nativeHeap * 10.0) / 10.0,
            label = "Native Heap (MB)"
        )
    } catch (_: Exception) {}

    // Memory allocation rate
    if (lastMemoryMB > 0) {
        val deltaMB = usedMB - lastMemoryMB
        val ratePerSec = Math.round(deltaMB / 5.0 * 100) / 100.0 // 5s interval
        DevConnect.reportPerformanceMetric(
            metricType = "memory_allocation_rate",
            value = ratePerSec,
            label = "${if (ratePerSec >= 0) "+" else ""}$ratePerSec MB/s"
        )
    }
    lastMemoryMB = usedMB
}

// ---- System metrics ----

private fun reportSystemMetrics() {
    // Thread count
    try {
        val threadCount = Thread.activeCount()
        DevConnect.reportPerformanceMetric(
            metricType = "thread_count",
            value = threadCount.toDouble(),
            label = "$threadCount threads"
        )
    } catch (_: Exception) {}

    // Disk I/O via /proc/self/io
    try {
        val io = java.io.File("/proc/self/io").readText()
        val readBytes = Regex("read_bytes:\\s+(\\d+)").find(io)?.groupValues?.get(1)?.toLongOrNull()
        val writeBytes = Regex("write_bytes:\\s+(\\d+)").find(io)?.groupValues?.get(1)?.toLongOrNull()
        if (readBytes != null) {
            val mb = Math.round(readBytes.toDouble() / 1024 / 1024 * 10) / 10.0
            DevConnect.reportPerformanceMetric(
                metricType = "disk_read",
                value = mb,
                label = "Disk Read: $mb MB"
            )
        }
        if (writeBytes != null) {
            val mb = Math.round(writeBytes.toDouble() / 1024 / 1024 * 10) / 10.0
            DevConnect.reportPerformanceMetric(
                metricType = "disk_write",
                value = mb,
                label = "Disk Write: $mb MB"
            )
        }
    } catch (_: Exception) {}

    // Battery level
    val ctx = appContext
    if (ctx != null) {
        try {
            val batteryIntent = ctx.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
            if (batteryIntent != null) {
                val level = batteryIntent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
                val scale = batteryIntent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
                val status = batteryIntent.getIntExtra(BatteryManager.EXTRA_STATUS, -1)
                val temp = batteryIntent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, -1)
                val isCharging = status == BatteryManager.BATTERY_STATUS_CHARGING || status == BatteryManager.BATTERY_STATUS_FULL
                if (level >= 0 && scale > 0) {
                    val pct = level * 100 / scale
                    DevConnect.reportPerformanceMetric(
                        metricType = "battery_level",
                        value = pct.toDouble(),
                        label = "Battery: $pct%${if (isCharging) " (charging)" else ""}",
                        metadata = mapOf(
                            "charging" to isCharging,
                            "temperature" to (temp / 10.0)
                        )
                    )
                }
            }
        } catch (_: Exception) {}
    }

    // Thermal state via thermal_zone0
    try {
        val temp = java.io.File("/sys/class/thermal/thermal_zone0/temp").readText().trim().toLongOrNull()
        if (temp != null) {
            val tempC = temp / 1000.0
            val state = when {
                tempC < 35 -> 0.0  // nominal
                tempC < 40 -> 1.0  // fair
                tempC < 45 -> 2.0  // serious
                else -> 3.0        // critical
            }
            DevConnect.reportPerformanceMetric(
                metricType = "thermal_state",
                value = state,
                label = "Thermal: ${Math.round(tempC * 10) / 10.0}°C",
                metadata = mapOf("temperatureC" to tempC)
            )
        }
    } catch (_: Exception) {}

    // ANR detection (main thread responsiveness)
    detectAnr()
}

// ---- ANR detection ----
private val mainThreadAck = java.util.concurrent.atomic.AtomicBoolean(false)

private fun detectAnr() {
    // The previous implementation reset `anrCheckTime` both on the
    // worker thread (before sleeping) and on the main thread (via
    // handler.post). That meant the worker's `delay = now - anrCheckTime`
    // was always ~5000 ms regardless of whether the main thread was
    // responsive, so the `> 6000` threshold was never reached — ANRs
    // were silently dropped.

    mainThreadAck.set(false)
    val handler = Handler(Looper.getMainLooper())
    handler.post { mainThreadAck.set(true) }

    Thread {
        Thread.sleep(5000)
        if (!running) return@Thread
        if (mainThreadAck.get()) return@Thread

        // Main thread didn't ack within 5 s. Sleep one more second to
        // rule out transient jank (GC pause, layout inflation, ...) and
        // re-check. Only report if it still hasn't responded.
        Thread.sleep(1000)
        if (!running) return@Thread
        if (mainThreadAck.get()) return@Thread

        // Capture the main thread's stack trace — we're on a background
        // thread right now, so reading `Looper.getMainLooper().thread`
        // is safe.
        val mainStack = Looper.getMainLooper().thread.stackTrace
            .take(20)
            .joinToString("\n") {
                "${it.className}.${it.methodName}(${it.fileName}:${it.lineNumber})"
            }

        DevConnect.reportPerformanceMetric(
            metricType = "anr",
            value = 6000.0,
            label = "ANR detected: main thread blocked ≥6s",
            metadata = mapOf(
                "blockDurationMs" to 6000,
                "mainThreadStack" to mainStack
            )
        )
    }.start()
}

File: client_sdks/devconnect_flutter/lib/src/plugins/performance_monitor.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/scheduler.dart';
import 'package:flutter/widgets.dart';

import '../devconnect_client.dart';

bool _running = false;
Timer? _memoryTimer;
Timer? _cpuTimer;
Timer? _systemTimer;
bool _startupReported = false;

class PerformanceMonitorOptions {
  final int fpsInterval;
  final int memoryInterval;
  final int cpuInterval;
  final int systemInterval;
  final double jankThresholdMs;

  const PerformanceMonitorOptions({
    this.fpsInterval = 2000,
    this.memoryInterval = 5000,
    this.cpuInterval = 3000,
    this.systemInterval = 10000,
    this.jankThresholdMs = 32.0,
  });
}

final _initTime = DateTime.now();

void startPerformanceMonitor([PerformanceMonitorOptions opts = const PerformanceMonitorOptions()]) {
  if (_running) return;
  _running = true;

  // Wait for 3 frames to ensure layout is stable before tracking
  WidgetsBinding.instance.addPostFrameCallback((_) {
    WidgetsBinding.instance.addPostFrameCallback((_) {
      WidgetsBinding.instance.addPostFrameCallback((_) {
        if (!_running) return;

        // ---- Startup Time ----
        if (!_startupReported) {
          final startupMs = DateTime.now().difference(_initTime).inMilliseconds;
          DevConnectClient.safeSend('client:performance:metric', {
            'metricType': 'startup_time',
            'value': startupMs.toDouble(),
            'label': 'App startup: ${startupMs}ms',
          });
          _startupReported = true;
        }

        // ---- Frame timing (FPS + jank + build/raster time) ----
        WidgetsBinding.instance.addTimingsCallback(_onFrameTimings);

        // ---- Memory monitor ----
        _memoryTimer = Timer.periodic(Duration(milliseconds: opts.memoryInterval), (_) {
          if (!_running) return;
          _reportMemory();
        });

        // ---- CPU monitor ----
        _cpuTimer = Timer.periodic(Duration(milliseconds: opts.cpuInterval), (_) {
          if (!_running) return;
          _reportCpu(opts.cpuInterval);
        });

        // ---- System metrics (battery, threads, disk) ----
        _reportSystemMetrics(); // Report immediately once
        _systemTimer = Timer.periodic(Duration(milliseconds: opts.systemInterval), (_) {
          if (!_running) return;
          _reportSystemMetrics();
        });
      });
    });
  });
}

void stopPerformanceMonitor() {
  _running = false;
  WidgetsBinding.instance.removeTimingsCallback(_onFrameTimings);
  _memoryTimer?.cancel();
  _memoryTimer = null;
  _cpuTimer?.cancel();
  _cpuTimer = null;
  _systemTimer?.cancel();
  _systemTimer = null;
  _totalBuildMs = 0;
  _totalRasterMs = 0;
  _cpuFrameCount = 0;
}

// ---- CPU estimation state ----
double _totalBuildMs = 0;
double _totalRasterMs = 0;
int _cpuFrameCount = 0;

// ---- Frame timing callback ----
int _frameCount = 0;
int _lastFpsReport = 0;
const _fpsReportInterval = 2000; // ms

void _onFrameTimings(List<FrameTiming> timings) {
  if (!_running) return;

  final now = DateTime.now().millisecondsSinceEpoch;
  if (_lastFpsReport == 0) _lastFpsReport = now;

  for (final timing in timings) {
    _frameCount++;

    final buildMs = timing.buildDuration.inMicroseconds / 1000.0;
    final rasterMs = timing.rasterDuration.inMicroseconds / 1000.0;
    final totalMs = timing.totalSpan.inMicroseconds / 1000.0;

    // Accumulate for CPU estimation
    _totalBuildMs += buildMs;
    _totalRasterMs += rasterMs;
    _cpuFrameCount++;

    // Report frame build time
    DevConnectClient.safeSend('client:performance:metric', {
      'metricType': 'frame_build_time',
      'value': (buildMs * 10).roundToDouble() / 10,
      'label': 'Build: ${buildMs.round()}ms',
      'metadata': {'rasterMs': rasterMs, 'totalMs': totalMs},
    });

    // Report frame raster time
    DevConnectClient.safeSend('client:performance:metric', {
      'metricType': 'frame_raster_time',
      'value': (rasterMs * 10).roundToDouble() / 10,
      'label': 'Raster: ${rasterMs.round()}ms',
    });

    // Detect jank
    if (totalMs > 32.0) {
      DevConnectClient.safeSend('client:performance:metric', {
        'metricType': 'jank_frame',
        'value': (totalMs * 10).roundToDouble() / 10,
        'label': 'Slow frame: ${totalMs.round()}ms (build: ${buildMs.round()}ms, raster: ${rasterMs.round()}ms)',
        'metadata': {
          'buildDuration': buildMs,
          'rasterDuration': rasterMs,
        },
      });
    }
  }

  // Report FPS periodically
  final elapsed = now - _lastFpsReport;
  if (elapsed >= _fpsReportInterval) {
    final fps = (_frameCount / elapsed * 1000 * 10).roundToDouble() / 10;
    DevConnectClient.safeSend('client:performance:metric', {
      'metricType': 'fps',
      'value': fps,
      'label': 'UI Thread FPS',
    });
    _frameCount = 0;
    _lastFpsReport = now;
  }
}

void _reportCpu(int intervalMs) {
  if (_cpuFrameCount == 0) return;
  final totalWorkMs = _totalBuildMs + _totalRasterMs;
  final usage = (totalWorkMs / intervalMs * 100).clamp(0.0, 100.0);
  DevConnectClient.safeSend('client:performance:metric', {
    'metricType': 'cpu_usage',
    'value': (usage * 10).roundToDouble() / 10,
    'label': 'UI Thread Utilization (%)',
    'metadata': {
      'buildTimeMs': (_totalBuildMs * 10).roundToDouble() / 10,
      'rasterTimeMs': (_totalRasterMs * 10).roundToDouble() / 10,
      'frames': _cpuFrameCount,
    },
  });
  _totalBuildMs = 0;
  _totalRasterMs = 0;
  _cpuFrameCount = 0;
}

// ---- Memory ----
double _lastMemoryMB = 0;

void _reportMemory() {
  try {
    double memoryMB = 0;
    String label = 'Memory (MB)';
    Map<String, dynamic> metadata = {};

    final rss = ProcessInfo.currentRss;
    if (rss > 0) {
      memoryMB = (rss / 1024 / 1024 * 10).roundToDouble() / 10;
      label = 'RSS Memory (MB)';
      metadata = {'maxRss': ProcessInfo.maxRss, 'rssBytes': rss};
    }

    if (memoryMB == 0 && Platform.isAndroid) {
      try {
        final status = File('/proc/self/status').readAsStringSync();
        final vmRss = RegExp(r'VmRSS:\s+(\d+)\s+kB').firstMatch(status);
        if (vmRss != null) {
          final kB = int.parse(vmRss.group(1)!);
          memoryMB = (kB / 1024 * 10).roundToDouble() / 10;
          label = 'VmRSS (MB)';
          final vmSize = RegExp(r'VmSize:\s+(\d+)\s+kB').firstMatch(status);
          metadata = {
            'vmRssKB': kB,
            if (vmSize != null) 'vmSizeKB': int.parse(vmSize.group(1)!),
          };
        }
      } catch (_) {}
    }

    if (memoryMB == 0) {
      final maxRss = ProcessInfo.maxRss;
      if (maxRss > 0) {
        memoryMB = (maxRss / 1024 / 1024 * 10).roundToDouble() / 10;
        label = 'Peak RSS (MB)';
        metadata = {'maxRssBytes': maxRss};
      }
    }

    if (memoryMB > 0) {
      DevConnectClient.safeSend('client:performance:metric', {
        'metricType': 'memory_usage',
        'value': memoryMB,
        'label': label,
        'metadata': metadata,
      });

      // Memory allocation rate
      if (_lastMemoryMB > 0) {
        final deltaMB = memoryMB - _lastMemoryMB;
        final ratePerSec = (deltaMB / 5.0 * 100).roundToDouble() / 100; // 5s interval
        DevConnectClient.safeSend('client:performance:metric', {
          'metricType': 'memory_allocation_rate',
          'value': ratePerSec,
          'label': '${ratePerSec >= 0 ? '+' : ''}$ratePerSec MB/s',
        });
      }
      _lastMemoryMB = memoryMB;
    }
  } catch (_) {}
}

// ---- System metrics ----
void _reportSystemMetrics() {
  // Thread count (Android via /proc/self/status)
  if (Platform.isAndroid) {
    try {
      final status = File('/proc/self/status').readAsStringSync();
      final threads = RegExp(r'Threads:\s+(\d+)').firstMatch(status);
      if (threads != null) {
        DevConnectClient.safeSend('client:performance:metric', {
          'metricType': 'thread_count',
          'value': int.parse(threads.group(1)!).toDouble(),
          'label': '${threads.group(1)} threads',
        });
      }
    } catch (_) {}

    // Disk I/O (Android via /proc/self/io)
    try {
      final io = File('/proc/self/io').readAsStringSync();
      final readBytes = RegExp(r'read_bytes:\s+(\d+)').firstMatch(io);
      final writeBytes = RegExp(r'write_bytes:\s+(\d+)').firstMatch(io);
      if (readBytes != null) {
        final mb = (int.parse(readBytes.group(1)!) / 1024 / 1024 * 10).roundToDouble() / 10;
        DevConnectClient.safeSend('client:performance:metric', {
          'metricType': 'disk_read',
          'value': mb,
          'label': 'Disk Read: $mb MB',
        });
      }
      if (writeBytes != null) {
        final mb = (int.parse(writeBytes.group(1)!) / 1024 / 1024 * 10).roundToDouble() / 10;
        DevConnectClient.safeSend('client:performance:metric', {
          'metricType': 'disk_write',
          'value': mb,
          'label': 'Disk Write: $mb MB',
        });
      }
    } catch (_) {}

    // Battery level (Android via /sys/class/power_supply/battery/)
    try {
      final capacity = File('/sys/class/power_supply/battery/capacity').readAsStringSync().trim();
      final status = File('/sys/class/power_supply/battery/status').readAsStringSync().trim();
      final level = int.tryParse(capacity);
      if (level != null) {
        DevConnectClient.safeSend('client:performance:metric', {
          'metricType': 'battery_level',
          'value': level.toDouble(),
          'label': 'Battery: $level% ($status)',
          'metadata': {'status': status},
        });
      }
    } catch (_) {}

    // Thermal (Android via thermal_zone0)
    try {
      final temp = File('/sys/class/thermal/thermal_zone0/temp').readAsStringSync().trim();
      final tempC = (int.tryParse(temp) ?? 0) / 1000.0;
      if (tempC > 0) {
        // 0=nominal, 1=fair, 2=serious, 3=critical
        final state = tempC < 35 ? 0.0 : tempC < 40 ? 1.0 : tempC < 45 ? 2.0 : 3.0;
        DevConnectClient.safeSend('client:performance:metric', {
          'metricType': 'thermal_state',
          'value': state,
          'label': 'Thermal: ${tempC.toStringAsFixed(1)}°C',
          'metadata': {'temperatureC': tempC},
        });
      }
    } catch (_) {}
  }

  // iOS — limited access from Dart, report what's available
  if (Platform.isIOS) {
    // Thread count via Isolate (main + other)
    DevConnectClient.safeSend('client:performance:metric', {
      'metricType': 'thread_count',
      'value': Platform.numberOfProcessors.toDouble(),
      'label': '${Platform.numberOfProcessors} cores',
      'metadata': {'note': 'CPU cores (thread count not available on iOS from Dart)'},
    });
  }
}
